### PR TITLE
New use case: getting route shapes (back-end)

### PR DESCRIPTION
**Pull request by Swithord**
The endpoint /shapes will return a mapping of route id/tags to a list of coordinates.

### CODE REVIEW

File: backend/src/main/java/com/example/backend/data_access/route/RouteDAO.java [hunk 1]
```diff
@@ -180,4 +186,43 @@ public HashMap<String, HashSet<String>> getStopTagsToRouteTags() {
...
+        // Used BufferedReader for massively improved performance
+        try (BufferedReader reader = new BufferedReader(new FileReader(shapesCsvFilename))) {
+            String line;
+            // Skip first row (headers)
+            reader.readLine();
+
+            while ((line = reader.readLine()) != null) {
+                String[] row = line.split(",");
+                // Get the routeId, latitude and longitude from the row
+                String routeId = row[1];
```
Comment (nsaroiu) [inline backend/src/main/java/com/example/backend/data_access/route/RouteDAO.java:202]: I'm confused. In the csv file, row[1] is marked as shape id. However, the variable is called route id. Which is it?

**Review by nsaroiu — CHANGES_REQUESTED** [review]
One note with the csv file, is there any way to remove the first column? It would save space (and also be less complex) if it were removed.

Other than that, the new method in RouteDAO needs to be polished up just a little bit.

File: backend/src/main/java/com/example/backend/data_access/route/RouteDAO.java (hunk 1 above)
Comment (Swithord) [reply backend/src/main/java/com/example/backend/data_access/route/RouteDAO.java:202]: It's supposed to be route id, the shape id column on the csv is a remnant from the original dataset which I forgot to change

File: backend/src/main/java/com/example/backend/data_access/route/RouteDAO.java [hunk 2]
```diff
@@ -180,4 +186,43 @@ public HashMap<String, HashSet<String>> getStopTagsToRouteTags() {
         return stopTagsToRouteTags;
     }
 
+    @Override
```
[the remark this answers is not in the document]
Comment (Swithord) [reply backend/src/main/java/com/example/backend/data_access/route/RouteDAO.java:189]: yep

File: backend/src/main/java/com/example/backend/data_access/route/RouteDAO.java [hunk 3]
```diff
@@ -180,4 +186,43 @@ public HashMap<String, HashSet<String>> getStopTagsToRouteTags() {
...
+    public HashMap<String, ArrayList<Location>> getRouteShapes() {
+        HashMap<String, ArrayList<Location>> routeShapes = new HashMap<>();
+
+        // Used BufferedReader for massively improved performance
+        try (BufferedReader reader = new BufferedReader(new FileReader(shapesCsvFilename))) {
```
[the remark this answers is not in the document]
Comment (Swithord) [reply backend/src/main/java/com/example/backend/data_access/route/RouteDAO.java:194]: it's already caught as FileNotFoundException is a subclass of IOException

**Comment by Swithord** [comment]
> One note with the csv file, is there any way to remove the first column? It would save space (and also be less complex) if it were removed.

I think it doesn't hurt to keep it for future reference for which column means
stopData.csv also has headers anyway

File: backend/src/main/java/com/example/backend/data_access/route/RouteDAO.java [hunk 4]
```diff
@@ -180,4 +186,43 @@ public HashMap<String, HashSet<String>> getStopTagsToRouteTags() {
         return stopTagsToRouteTags;
     }
```
[the remark this answers is not in the document]
Comment (Swithord) [reply backend/src/main/java/com/example/backend/data_access/route/RouteDAO.java:188]: 👍

**Review by nsaroiu — APPROVED** [review]
lgtm lol

For the csv file, I'm not entirely sure what you're trying to say? I was arguing that it might be a good idea in the future to remove the first column (not the first/header row) because it just seems to be a bunch of incrementing numbers. For now though, the code works and this is a super minor issue, so I'm going to approve the merge.

**Comment by Swithord** [comment]
> lgtm lol
> 
> For the csv file, I'm not entirely sure what you're trying to say? I was arguing that it might be a good idea in the future to remove the first column (not the first/header row) because it just seems to be a bunch of incrementing numbers. For now though, the code works and this is a super minor issue, so I'm going to approve the merge.

oh sorry, I didn't realise you were talking about the index column. it's probably better without it but it also doesn't really matter anyway

**Pull request merged**